### PR TITLE
Add tests for Bridge/PGO integration code that hits the Bridge API.

### PR DESCRIPTION
**Pull request by dsessler7**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change, as applicable?
 - [ ] Have you tested your changes on all related environments with successful results, as applicable?
   - [x] Have you added automated tests?

**Type of Changes:**

 
 - [ ] New feature
 - [ ] Bug fix
 - [ ] Documentation
 - [x] Testing enhancement
 - [ ] Other

**What is the current behavior (link to any open issues here)?**

Currently, for our CrunchyBridgeCluster controller code, we only have tests in place for the code that doesn't interact with the Bridge API.

**What is the new behavior (if this is a feature change)?**
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

This PR adds tests for the Bridge Client methods that send/receive HTTP requests/responses. It also adds a mock Bridge Client, complete with mock methods, that are used for testing the code that calls the aforementioned Bridge Client methods and uses their responses.

**Other Information**:

### CODE REVIEW

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go [hunk 1]
```diff
@@ -20,33 +20,961 @@ package crunchybridgecluster
...
+			TeamId:   "5678",
+			Clusters: []*bridge.ClusterApiResource{clusterInBridge},
+		}
+	}
+
+	t.Run("UpgradePlan", func(t *testing.T) {
+		cluster := testCluster()
+		cluster.Namespace = ns
+		cluster.Status.ID = "1234"
+		cluster.Spec.Plan = "standard-16"
```
Comment (jmckulk) [inline internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go:712]: Maybe comment that the existing plan is different? Something like...

```suggestion
		cluster.Spec.Plan = "standard-16" // Originally "standard-8"
```

Same for some of these other tests. It'd be nice to be able to see what is different from the `testCluster` without having to look into another file

File: internal/bridge/client_test.go [hunk 2]
```diff
@@ -536,3 +542,957 @@ func TestClientCreateInstallation(t *testing.T) {
...
+		}))
+		t.Cleanup(server.Close)
+
+		client := NewClient(server.URL, "")
+		assert.Equal(t, client.BaseURL.String(), server.URL)
+
+		cluster, err := client.GetCluster(context.Background(), apiKey, clusterId)
+		assert.NilError(t, err)
+		assert.Equal(t, cluster.ClusterName, clusterApiResource.ClusterName)
+		// TODO: check every field?
```
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/client_test.go:921]: I don't think we need to. Checking every field would really just be making sure that the `http` package properly sends payloads and that the `json` package properly unmarshals, which we don't need to be testing. I'll remove the TODOs.

File: internal/bridge/client_test.go [hunk 3]
```diff
@@ -536,3 +542,957 @@ func TestClientCreateInstallation(t *testing.T) {
...
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusOK)
+			_, _ = w.Write(responsePayloadJson)
+		}))
+		t.Cleanup(server.Close)
+
+		client := NewClient(server.URL, "")
+		assert.Equal(t, client.BaseURL.String(), server.URL)
+
+		clusterRole, err := client.GetClusterRole(context.Background(), apiKey, clusterId, roleName)
```
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/client_test.go:1386]: The test is written so that the http server sends a hard-coded response... In these tests, we aren't actually mocking the API, we are simply ensuring that we are sending the right data to Bridge (in these tests, a fake http server) and that when we get a 200/400 response, that we return the right things.

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go [hunk 4]
```diff
@@ -20,33 +20,961 @@ package crunchybridgecluster
...
+type TestBridgeClient struct {
+	ApiKey          string
+	TeamId          string
+	Clusters        []*bridge.ClusterApiResource
+	ClusterRoles    []*bridge.ClusterRoleApiResource
+	ClusterStatuses map[string]*bridge.ClusterStatusApiResource
+	ClusterUpgrades map[string]*bridge.ClusterUpgradeApiResource
+}
+
+func (tbc *TestBridgeClient) ListClusters(ctx context.Context, apiKey, teamId string) ([]*bridge.ClusterApiResource, error) {
```
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go:57]: I guess it depends on how much the Bridge folks change the API, but yes, these methods could require adjusting as the Bridge API changes...

Where do you think they belong? Maybe `helpers_test.go`? Or their own file? `mock_bridge_api.go`?

[the remark this answers is not in the document]
Comment (jmckulk) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go]: I could see a `mock_bridge_api.go` or similar

File: internal/bridge/client_test.go [hunk 5]
```diff
@@ -536,3 +542,957 @@ func TestClientCreateInstallation(t *testing.T) {
...
+func TestUpdateCluster(t *testing.T) {
+	clusterId := "1234"
+	apiKey := "9012"
+
+	clusterApiResource := &ClusterApiResource{
+		ClusterName:  "new-cluster-name",
+		Host:         "new-cluster-name.crunchybridge.com",
+		ID:           clusterId,
+		IsHA:         initialize.Bool(false),
+		IsProtected:  initialize.Bool(false),
```
[the remark this answers is not in the document]
Comment (jmckulk) [reply internal/bridge/client_test.go]: If the test wasn't failing with the difference, do we need it?

File: internal/bridge/client_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmckulk) [reply internal/bridge/client_test.go:1370]: hm - so `NewClient` here is using the mock functions that you just moved to the new file?

File: internal/bridge/crunchybridgecluster/postgres.go [hunk 6]
```diff
@@ -156,7 +156,7 @@ func (r *CrunchyBridgeClusterReconciler) reconcilePostgresRoleSecrets(
 		clusterRole, err := r.NewClient().GetClusterRole(ctx, apiKey, cluster.Status.ID, roleName)
 		// If issue with getting ClusterRole, log error and move on to next role
 		if err != nil {
-			// TODO (dsessler7): Emit event here?
+			// TODO(dsessler7): Emit event here?
```
Comment (jmckulk) [inline internal/bridge/crunchybridgecluster/postgres.go:159]: I think we would want this space to be consistent? Not a huge deal but if you end up needing more changes

File: internal/bridge/client_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/client_test.go]: No, but that can be said about many things in many places in these tests, and IMO it's not worth the time/effort to go back and take every piece of data that isn't necessary out... Why is it being set? Consistency, I guess.

File: internal/bridge/client_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/client_test.go:1370]: No, `NewClient` here returns a `bridge.Client` but it sets the `BaseURL` to the test http server's (defined above on L1361) URL so that requests get sent to the test http server rather than to the Bridge API... The mock bridge api is used by the tests in `crunchybridgecluster_controller_test.go`. They replace the `NewClient` function with one that returns a `TestBridgeClient` rather than a `bridge.Client`.

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller_test.go]: That's what I went with

File: internal/bridge/crunchybridgecluster/delete_test.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+		assert.Equal(t, *controllerResult, ctrl.Result{})
+
+		// Run handleDelete a couple times with good api key so test can cleanup properly
+		// delete from bridge
+		_, err = reconciler.handleDelete(ctx, cluster, "9012")
+		assert.NilError(t, err)
+
+		// remove finalizer
+		_, err = reconciler.handleDelete(ctx, cluster, "9012")
+		assert.NilError(t, err)
```
Comment (jmckulk) [inline internal/bridge/crunchybridgecluster/delete_test.go:135]: Why are we running this twice without any changes in between? Do we expect that multiple runs of `handleDelete` will do different things? 

I see that in the successful delete we do the same thing.

Comment (dsessler7) [reply internal/bridge/crunchybridgecluster/delete_test.go:135]: Yeah if you look at the `handleDelete` function, when a `DeletionTimestamp` is applied to the CR, the first time `handleDelete` is run it sends the request to Bridge to delete the cluster. The response *should* have `deletedAlready` as `false` so it requeues. When `handleDelete` gets run on the requeue loop, the request is made to Bridge again, and the response *should* now result in `deletedAlready` being `true`, so we then remove the finalizer... 

TL;DR: we don't want to remove the finalizer until we're sure that the cluster has been deleted from Bridge, so we have to do multiple calls/reconcile loops.

Comment (jmckulk) [reply internal/bridge/crunchybridgecluster/delete_test.go:135]: Ah, that makes sense. I think that TL;DR as a comment would be helpful

**Review by jmckulk — APPROVED** [review]
LGTM
